apps/emqx_bridge_kafka/src/emqx_bridge_kafka_impl_producer.erl

```diff
@@ -294,34 +302,102 @@ on_kafka_ack(_Partition, buffer_overflow_discarded, _Callback) ->
+        Status ->
+            Status
+    end.
+
+do_get_healthy_leaders(Client, KafkaTopic) ->
     case wolff_client:get_leader_connections(Client, KafkaTopic) of
         {ok, Leaders} ->
-            %% Kafka is considered healthy as long as any of the partition leader is reachable
-            case
-                lists:any(
-                    fun({_Partition, Pid}) ->
-                        is_pid(Pid) andalso erlang:is_process_alive(Pid)
-                    end,
-                    Leaders
-                )
-            of
-                true ->
-                    connected;
-                false ->
-                    connecting
-            end;
+            %% Kafka is considered healthy as long as any of the partition leader is reachable.
+            lists:filtermap(
+                fun({_Partition, Pid}) ->
+                    case is_pid(Pid) andalso erlang:is_process_alive(Pid) of
+                        true -> {true, Pid};
+                        _ -> false
+                    end
+                end,
+                Leaders
+            );
         {error, _} ->
-            connecting
+            []
+    end.
+
+do_get_topic_status(PidList, Topic) when is_list(PidList) ->
+    lists:foldl(
+        fun
+            (_Pid, ok) ->
+                ok;
+            (Pid, Acc) ->
+                case do_get_topic_status(Pid, Topic) of
+                    error -> Acc;
+                    Status -> Status
+                end
+        end,
+        error,
+        PidList
+    );
+do_get_topic_status(Pid, Topic) ->
+    case wolff_client:check_if_topic_exists(Pid, Topic) of
+        ok ->
+            ok;
+        {error, unknown_topic_or_partition} ->
+            unhealthy_target;
+        _ ->
+            error
+    end.
+
+do_get_topic_status(Hosts, KafkaConfig, KafkaTopic) ->
+    RetPid = self(),
+    QueryRef = make_ref(),
+    CheckTopic = fun() ->
+        Ret =
+            case wolff_client:check_if_topic_exists(Hosts, KafkaConfig, KafkaTopic) of
+                ok ->
+                    ok;
+                {error, unknown_topic_or_partition} ->
+                    unhealthy_target;
+                _ ->
+                    error
+            end,
+        RetPid ! {QueryRef, Ret}
+    end,
+    % wolff starts a socket to get metadata and closes it asynchronously (calling a spawn to
+    % actually close it), causing some {'EXIT', tcp_closed} errors. Therefore, it'll be executed
+    % by a spawn process to avoid crashing the main process.
+    {Pid, Ref} = spawn_monitor(CheckTopic),
+    % wait process terminate
+    receive
+        {'DOWN', Ref, process, Pid, _What} -> ok
+    after 5000 ->
+        ?SLOG(error, #{msg => "failed to check kafka topic", pid => Pid}),
+        demonitor(Ref),
+        exit(Pid, kill)
+    end,
+    % check for result
+    receive
+        {QueryRef, Result} -> Result
+    after 500 ->
+        error
```

Comment:
It seems like this could use `emqx_utils:nolink_apply`.

```erlang
try 
  emqx_utils:nolink_apply(fun() -> check... end, Timeout)
catch
  exit:_ -> ...;
  error:_ -> ...
```